### PR TITLE
add docs for finite difference operator internals

### PR DESCRIPTION
**Pull request by simonbyrne**
Document some of the internal APIs

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

File: src/Operators/finitedifference.jl [hunk 1]
```diff
@@ -969,7 +1044,6 @@ Supported boundary conditions are:
 struct UpwindBiasedProductC2F{BCS} <: AdvectionOperator
     bcs::BCS
 end
-UpwindBiasedProductC2F(; kwargs...) = UpwindBiasedProductC2F(NamedTuple(kwargs))
```
Comment (charleskawczynski) [inline src/Operators/finitedifference.jl:972]: Why is this being deleted?

Comment (charleskawczynski) [reply src/Operators/finitedifference.jl:972]: We use `UpwindBiasedProductC2F` in TC ([here](https://github.com/CliMA/TurbulenceConvection.jl/blob/10a7f477995da0c73ebb7535c990a85c17f1f680/src/EDMF_functions.jl#L452))

File: src/Operators/finitedifference.jl [hunk 2]
```diff
@@ -285,14 +307,67 @@ struct CompositeStencilStyle <: AbstractStencilStyle end
...
+implies that at index `i`, the stencil accesses `arg1` at `i-half`, `i+half` and
+`i+1+half`, and `arg2` at index `i`.
 """
 function stencil_interior_width end
 
+"""
+    stencil_interior(::Op, loc, idx, args...)
+
+Defines the stencil of the operator `Op` in the interior of the domain at `idx`,
+`args` are the input arguments.
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:344]: Can we please clarify this sentence? Is the last bit extra information? As in:

```suggestion
Defines the stencil of the operator `Op` in the interior of the domain at `idx`;
`args` are the input arguments.
```
?

And can we describe `loc` as well?

File: src/Operators/finitedifference.jl [hunk 3]
```diff
@@ -285,14 +307,67 @@ struct CompositeStencilStyle <: AbstractStencilStyle end
...
+"""
+function stencil_interior end
+
+
+"""
+    boundary_width(::Op, ::BC, args...)
+
+Defines the width of a boundary condition `BC` on an operator `Op`. This is the
+number of locations that a modified stencil is used.
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:353]:
```suggestion
number of locations that are used in a modified stencil.
```
Can we also give examples here as you did for the interior stencil?

File: src/Operators/finitedifference.jl [hunk 4]
```diff
@@ -285,14 +307,67 @@ struct CompositeStencilStyle <: AbstractStencilStyle end
...
+"""
+function boundary_width end
+
+"""
+    stencil_left_boundary(::Op, ::BC, loc, idx, args...)
+
+Defines the stencil of operator `Op` with at `idx` near the left boundary, with boundary condition `BC`.
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:360]:
```suggestion
Defines the stencil of operator `Op` at `idx` near the left boundary, with boundary condition `BC`.
```

File: src/Operators/finitedifference.jl [hunk 5]
```diff
@@ -285,14 +307,67 @@ struct CompositeStencilStyle <: AbstractStencilStyle end
...
+"""
+function stencil_left_boundary end
+
+"""
+    stencil_right_boundary(::Op, ::BC, loc, idx, args...)
+
+Defines the stencil of operator `Op` with at `idx` near the right boundary, with boundary condition `BC`.
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:367]:
```suggestion
Defines the stencil of operator `Op` at `idx` near the right boundary, with boundary condition `BC`.
```

File: src/Operators/finitedifference.jl (hunk 1 above)
Comment (simonbyrne) [reply src/Operators/finitedifference.jl:972]: Hmm, it shouldn't be...

File: src/Operators/finitedifference.jl [hunk 6]
```diff
@@ -285,14 +307,67 @@ struct CompositeStencilStyle <: AbstractStencilStyle end
...
+```
+stencil(::Op, arg1, arg2) = ((-half, 1+half), (0,0))
+```
+implies that at index `i`, the stencil accesses `arg1` at `i-half`, `i+half` and
+`i+1+half`, and `arg2` at index `i`.
 """
 function stencil_interior_width end
 
+"""
+    stencil_interior(::Op, loc, idx, args...)
```
Comment (valeriabarra) [inline src/Operators/finitedifference.jl:341]: Should we define `loc` as well? Also in the left/right boundary stencils

Comment (valeriabarra) [reply src/Operators/finitedifference.jl:341]: @simonbyrne not sure you saw my question
